Inital pull request to decoder

generall opened the pull request:
Hi, this is just a initial pull-request to signal the start of work.

I created some basic interfaces, here is the main thought:
Each part of the decoder that depends on Vocabulary should stay in Model class, including BeamSearch, sample scheduler, `_output_projection_layer`.

Decoder should implement same functionality as `_prepare_output_projections` implements now with some minor changes.

nelson-liu committed:
Fix multiGPU peak gpu memory test (#2254)

This test was failing on a multigpu machine i use:

```
$ py.test -k "trainer_test"
=============================================================================================== test session starts ================================================================================================
platform linux -- Python 3.6.5, pytest-3.6.2, py-1.5.3, pluggy-0.6.0
rootdir: /home/nfliu/Documents/Github/allennlp, inifile: pytest.ini
plugins: cov-2.5.1, flaky-3.4.0
collected 1161 items / 1140 deselected

allennlp/tests/training/trainer_test.py .............F.......                                                                                                                                                [100%]

===================================================================================================== FAILURES =====================================================================================================
__________________________________________________________________________________ TestTrainer.test_trainer_can_run_multiple_gpu ___________________________________________________________________________________

self = <allennlp.tests.training.trainer_test.TestTrainer testMethod=test_trainer_can_run_multiple_gpu>

    @pytest.mark.skipif(torch.cuda.device_count() < 2,
                        reason="Need multiple GPUs.")
    def test_trainer_can_run_multiple_gpu(self):

        class MetaDataCheckWrapper(Model):
            """
                Checks that the metadata field has been correctly split across the batch dimension
                when running on multiple gpus.
                """
            def __init__(self, model):
                super().__init__(model.vocab)
                self.model = model

            def forward(self, **kwargs) -> Dict[str, torch.Tensor]:  # type: ignore # pylint: disable=arguments-differ
                assert 'metadata' in kwargs and 'tags' in kwargs, \
                    f'tokens and metadata must be provided. Got {kwargs.keys()} instead.'
                batch_size = kwargs['tokens']['tokens'].size()[0]
                assert len(kwargs['metadata']) == batch_size, \
                    f'metadata must be split appropriately. Expected {batch_size} elements, ' \
                    f"got {len(kwargs['metadata'])} elements."
                return self.model.forward(**kwargs)

        multigpu_iterator = BasicIterator(batch_size=4)
        multigpu_iterator.index_with(self.vocab)
        trainer = Trainer(MetaDataCheckWrapper(self.model), self.optimizer,
                          multigpu_iterator, self.instances, num_epochs=2,
                          cuda_device=[0, 1])
        metrics = trainer.train()
        assert 'peak_cpu_memory_MB' in metrics
        assert isinstance(metrics['peak_cpu_memory_MB'], float)
        assert metrics['peak_cpu_memory_MB'] > 0
        assert 'peak_gpu_0_memory_MB' in metrics
>       assert isinstance(metrics['peak_gpu_0_memory_MB'], float)
E       AssertionError: assert False
E        +  where False = isinstance(549, float)

allennlp/tests/training/trainer_test.py:128: AssertionError
----------------------------------------------------------------------------------------------- Captured stderr call -----------------------------------------------------------------------------------------------
12/27/2018 09:34:40 - INFO - allennlp.common.checks -   Pytorch version: 0.4.1
0it [00:00, ?it/s]12/27/2018 09:34:40 - INFO - allennlp.data.dataset_readers.sequence_tagging -   Reading instances from lines in file at: /home/nfliu/Documents/Github/allennlp/allennlp/tests/fixtures/data/sequence_tagging.tsv
4it [00:00, 12409.18it/s]
12/27/2018 09:34:40 - INFO - allennlp.data.vocabulary -   Fitting token dictionary from dataset.
100%|██████████| 4/4 [00:00<00:00, 71392.41it/s]
12/27/2018 09:34:40 - INFO - allennlp.common.from_params -   instantiating class <class 'allennlp.models.simple_tagger.SimpleTagger'> from params {'text_field_embedder': {'token_embedders': {'tokens': {'type': 'embedding', 'embedding_dim': 5}}}, 'encoder': {'type': 'lstm', 'input_size': 5, 'hidden_size': 7, 'num_layers': 2}} and extras {'vocab': <allennlp.data.vocabulary.Vocabulary object at 0x7fe36c0f5438>}
12/27/2018 09:34:40 - INFO - allennlp.common.from_params -   instantiating class <class 'allennlp.modules.text_field_embedders.text_field_embedder.TextFieldEmbedder'> from params {'token_embedders': {'tokens': {'type': 'embedding', 'embedding_dim': 5}}} and extras {'vocab': <allennlp.data.vocabulary.Vocabulary object at 0x7fe36c0f5438>}
12/27/2018 09:34:40 - INFO - allennlp.common.from_params -   instantiating class <class 'allennlp.modules.token_embedders.token_embedder.TokenEmbedder'> from params {'type': 'embedding', 'embedding_dim': 5} and extras {'vocab': <allennlp.data.vocabulary.Vocabulary object at 0x7fe36c0f5438>}
12/27/2018 09:34:40 - INFO - allennlp.common.from_params -   instantiating class <class 'allennlp.modules.seq2seq_encoders.seq2seq_encoder.Seq2SeqEncoder'> from params {'type': 'lstm', 'input_size': 5, 'hidden_size': 7, 'num_layers': 2} and extras {'vocab': <allennlp.data.vocabulary.Vocabulary object at 0x7fe36c0f5438>}
12/27/2018 09:34:40 - WARNING - allennlp.training.trainer -   Multiple GPU support is experimental not recommended for use. In some cases it may lead to incorrect results or undefined behavior.
12/27/2018 09:34:40 - INFO - allennlp.training.trainer -   Beginning training.
12/27/2018 09:34:40 - INFO - allennlp.training.trainer -   Epoch 0/1
12/27/2018 09:34:40 - INFO - allennlp.training.trainer -   Peak CPU memory usage MB: 2121.256
12/27/2018 09:34:40 - INFO - allennlp.training.trainer -   GPU 0 memory usage MB: 533
12/27/2018 09:34:40 - INFO - allennlp.training.trainer -   GPU 1 memory usage MB: 10
12/27/2018 09:34:40 - INFO - allennlp.training.trainer -   Training
  0%|          | 0/1 [00:00<?, ?it/s]12/27/2018 09:34:40 - DEBUG - allennlp.data.iterators.data_iterator -   Batch padding lengths: {'tokens': {'tokens_length': 4, 'num_tokens': 4}, 'tags': {'num_tokens': 4}}
12/27/2018 09:34:40 - DEBUG - allennlp.data.iterators.data_iterator -   Batch size: 4
loss: 0.7543 ||: 100%|██████████| 1/1 [00:02<00:00,  2.04s/it]
12/27/2018 09:34:42 - INFO - allennlp.training.trainer -                       Training |  Validation
12/27/2018 09:34:42 - INFO - allennlp.training.trainer -   loss            |     0.754  |       N/A
12/27/2018 09:34:42 - INFO - allennlp.training.trainer -   gpu_1_memory_MB |    10.000  |       N/A
12/27/2018 09:34:42 - INFO - allennlp.training.trainer -   gpu_0_memory_MB |   533.000  |       N/A
12/27/2018 09:34:42 - INFO - allennlp.training.trainer -   cpu_memory_MB   |  2121.256  |       N/A
12/27/2018 09:34:42 - INFO - allennlp.training.trainer -   Epoch duration: 00:00:02
12/27/2018 09:34:42 - INFO - allennlp.training.trainer -   Estimated training time remaining: 0:00:02
12/27/2018 09:34:42 - INFO - allennlp.training.trainer -   Epoch 1/1
12/27/2018 09:34:42 - INFO - allennlp.training.trainer -   Peak CPU memory usage MB: 3892.764
12/27/2018 09:34:42 - INFO - allennlp.training.trainer -   GPU 0 memory usage MB: 549
12/27/2018 09:34:42 - INFO - allennlp.training.trainer -   GPU 1 memory usage MB: 549
12/27/2018 09:34:42 - INFO - allennlp.training.trainer -   Training
  0%|          | 0/1 [00:00<?, ?it/s]12/27/2018 09:34:42 - DEBUG - allennlp.data.iterators.data_iterator -   Batch padding lengths: {'tokens': {'tokens_length': 4, 'num_tokens': 4}, 'tags': {'num_tokens': 4}}
12/27/2018 09:34:42 - DEBUG - allennlp.data.iterators.data_iterator -   Batch size: 4
loss: 0.7522 ||: 100%|██████████| 1/1 [00:00<00:00, 59.00it/s]
12/27/2018 09:34:42 - INFO - allennlp.training.trainer -                       Training |  Validation
12/27/2018 09:34:42 - INFO - allennlp.training.trainer -   loss            |     0.752  |       N/A
12/27/2018 09:34:42 - INFO - allennlp.training.trainer -   gpu_1_memory_MB |   549.000  |       N/A
12/27/2018 09:34:42 - INFO - allennlp.training.trainer -   gpu_0_memory_MB |   549.000  |       N/A
12/27/2018 09:34:42 - INFO - allennlp.training.trainer -   cpu_memory_MB   |  3892.764  |       N/A
12/27/2018 09:34:42 - INFO - allennlp.training.trainer -   Epoch duration: 00:00:00
------------------------------------------------------------------------------------------------ Captured log call -------------------------------------------------------------------------------------------------
09:34:40 - INFO - allennlp.common.checks - Pytorch version: 0.4.1
09:34:40 - INFO - allennlp.data.dataset_readers.sequence_tagging - Reading instances from lines in file at: /home/nfliu/Documents/Github/allennlp/allennlp/tests/fixtures/data/sequence_tagging.tsv
09:34:40 - INFO - allennlp.data.vocabulary - Fitting token dictionary from dataset.
09:34:40 - INFO - allennlp.common.from_params - instantiating class <class 'allennlp.models.simple_tagger.SimpleTagger'> from params {'text_field_embedder': {'token_embedders': {'tokens': {'type': 'embedding', 'embedding_dim': 5}}}, 'encoder': {'type': 'lstm', 'input_size': 5, 'hidden_size': 7, 'num_layers': 2}} and extras {'vocab': <allennlp.data.vocabulary.Vocabulary object at 0x7fe36c0f5438>}
09:34:40 - INFO - allennlp.common.from_params - instantiating class <class 'allennlp.modules.text_field_embedders.text_field_embedder.TextFieldEmbedder'> from params {'token_embedders': {'tokens': {'type': 'embedding', 'embedding_dim': 5}}} and extras {'vocab': <allennlp.data.vocabulary.Vocabulary object at 0x7fe36c0f5438>}
09:34:40 - INFO - allennlp.common.from_params - instantiating class <class 'allennlp.modules.token_embedders.token_embedder.TokenEmbedder'> from params {'type': 'embedding', 'embedding_dim': 5} and extras {'vocab': <allennlp.data.vocabulary.Vocabulary object at 0x7fe36c0f5438>}
09:34:40 - INFO - allennlp.common.from_params - instantiating class <class 'allennlp.modules.seq2seq_encoders.seq2seq_encoder.Seq2SeqEncoder'> from params {'type': 'lstm', 'input_size': 5, 'hidden_size': 7, 'num_layers': 2} and extras {'vocab': <allennlp.data.vocabulary.Vocabulary object at 0x7fe36c0f5438>}
09:34:40 - WARNING - allennlp.training.trainer - Multiple GPU support is experimental not recommended for use. In some cases it may lead to incorrect results or undefined behavior.
09:34:40 - INFO - allennlp.training.trainer - Beginning training.
09:34:40 - INFO - allennlp.training.trainer - Epoch 0/1
09:34:40 - INFO - allennlp.training.trainer - Peak CPU memory usage MB: 2121.256
09:34:40 - INFO - allennlp.training.trainer - GPU 0 memory usage MB: 533
09:34:40 - INFO - allennlp.training.trainer - GPU 1 memory usage MB: 10
09:34:40 - INFO - allennlp.training.trainer - Training
09:34:40 - DEBUG - allennlp.data.iterators.data_iterator - Batch padding lengths: {'tokens': {'tokens_length': 4, 'num_tokens': 4}, 'tags': {'num_tokens': 4}}
09:34:40 - DEBUG - allennlp.data.iterators.data_iterator - Batch size: 4
09:34:42 - INFO - allennlp.training.trainer -                     Training |  Validation
09:34:42 - INFO - allennlp.training.trainer - loss            |     0.754  |       N/A
09:34:42 - INFO - allennlp.training.trainer - gpu_1_memory_MB |    10.000  |       N/A
09:34:42 - INFO - allennlp.training.trainer - gpu_0_memory_MB |   533.000  |       N/A
09:34:42 - INFO - allennlp.training.trainer - cpu_memory_MB   |  2121.256  |       N/A
09:34:42 - INFO - allennlp.training.trainer - Epoch duration: 00:00:02
09:34:42 - INFO - allennlp.training.trainer - Estimated training time remaining: 0:00:02
09:34:42 - INFO - allennlp.training.trainer - Epoch 1/1
09:34:42 - INFO - allennlp.training.trainer - Peak CPU memory usage MB: 3892.764
09:34:42 - INFO - allennlp.training.trainer - GPU 0 memory usage MB: 549
09:34:42 - INFO - allennlp.training.trainer - GPU 1 memory usage MB: 549
09:34:42 - INFO - allennlp.training.trainer - Training
09:34:42 - DEBUG - allennlp.data.iterators.data_iterator - Batch padding lengths: {'tokens': {'tokens_length': 4, 'num_tokens': 4}, 'tags': {'num_tokens': 4}}
09:34:42 - DEBUG - allennlp.data.iterators.data_iterator - Batch size: 4
09:34:42 - INFO - allennlp.training.trainer -                     Training |  Validation
09:34:42 - INFO - allennlp.training.trainer - loss            |     0.752  |       N/A
09:34:42 - INFO - allennlp.training.trainer - gpu_1_memory_MB |   549.000  |       N/A
09:34:42 - INFO - allennlp.training.trainer - gpu_0_memory_MB |   549.000  |       N/A
09:34:42 - INFO - allennlp.training.trainer - cpu_memory_MB   |  3892.764  |       N/A
09:34:42 - INFO - allennlp.training.trainer - Epoch duration: 00:00:00
===Flaky Test Report===


===End Flaky Test Report===
================================================================================================= warnings summary =================================================================================================
allennlp/tests/training/trainer_test.py::TestTrainer::test_trainer_can_run_multiple_gpu
  /home/nfliu/Documents/Github/allennlp/allennlp/modules/encoder_base.py:116: UserWarning: RNN module weights are not part of single contiguous chunk of memory. This means they need to be compacted at every call, possibly greatly increasing memory usage. To compact weights again call flatten_parameters().
    module_output, final_states = module(packed_sequence_input, initial_states)
  /home/nfliu/Documents/Github/allennlp/allennlp/modules/encoder_base.py:116: UserWarning: RNN module weights are not part of single contiguous chunk of memory. This means they need to be compacted at every call, possibly greatly increasing memory usage. To compact weights again call flatten_parameters().
    module_output, final_states = module(packed_sequence_input, initial_states)

-- Docs: http://doc.pytest.org/en/latest/warnings.html
======================================================================== 1 failed, 20 passed, 1140 deselected, 2 warnings in 26.92 seconds =========================================================================
```

Pretty sure it should be checking for `int` instead of `float`, since `common.util.gpu_memory_mb()` returns `Dict[int, int]`. See https://github.com/allenai/allennlp/blob/3f0953d19de3676ea82e642659fc96d90690e34d/allennlp/common/util.py#L327

ShomyLiu committed:
clarify the "num_output_representations" more clear (#2256)

brendan-ai2 committed:
Add a quick README for training transformer ELMo. (#2231)

- Training instructions.
- Example of using the trained model as an embedder.

seal6363 committed:
fix doc (#2213)

nelson-liu committed:
Deprecate bidirectional-language-model for bidirectional_language_model (#2253)

Commit:
Fix bug in uniform_unit_scaling #2239 (#2273)

pmandera committed:
Fix type annotation for .forward(...) in tutorial (#2122)

schmmd committed:
Add a Contributions section to README.md (#2277)

joelgrus committed:
script for doing archive surgery (#2223)

* script for doing archive surgery

* simplify script

nelson-liu committed:
Fix spelling in tutorial README (#2283)

WrRan committed:
fix #2285 (#2286)

WrRan committed:
Update the `find-lr` subcommand help text. (#2289)

* Update the elmo command help text.

* Update the find-lr subcommand help text.

nelson-liu committed:
Add __repr__ to Vocabulary (#2293)

As it currently stands, the following is logged during training:

```
2019-01-06 10:46:21,832 - INFO - allennlp.common.from_params - instantiating class <class 'allennlp.model
s.language_model.LanguageModel'> from params {'bidirectional': False, 'contextualizer': {'bidirectional':
 False, 'dropout': 0.5, 'hidden_size': 200, 'input_size': 200, 'num_layers': 2, 'type': 'lstm'}, 'dropout
': 0.5, 'text_field_embedder': {'token_embedders': {'tokens': {'embedding_dim': 200, 'type': 'embedding'}
}}} and extras {'vocab': <allennlp.data.vocabulary.Vocabulary object at 0x7ff7811665f8>}
```

Note that the `Vocabulary` does not provide any useful information, since it doesn't have `__repr__` defined. This provides a fix.

schmmd committed:
Update the base image in the Dockerfiles. (#2298)

nelson-liu committed:
Don't deprecate bidirectional-language-model name (#2297)

schmmd committed:
bump version number to v0.8.1

schmmd committed:
Bump version numbers to v0.8.2-unreleased

nelson-liu committed:
Turn BidirectionalLM into a more-general LanguageModel class (#2264)

Fixes https://github.com/allenai/allennlp/issues/2255

This PR replaces the `BidirectionalLM` class with a more-general `LanguageModel` that can be used in either the unidirectional/forward setting or the bidirectional setting.

It also accordingly replaces the `BidirectionalLanguageModelTokenEmbedder` with a `LanguageModelTokenEmbedder`.

Also fixes bug in the experiment_unsampled.jsonnet config that was preventing a test from actually being unsampled.

TODO:

- [x] test the unidirectional case
- [x] properly deprecate `BidirectionalLM` and `BidirectionalLanguageModelTokenEmbedder` 
- [x] check docs for accuracy
- [x] fix user-facing training configs

matt-gardner committed:
Grammar induction from a python executor (#2281)

* Got a simple LF executor working where all you need is to specify functions

* Grammar induction for simple cases now works

* Got logical_form_to_action_sequence working for simple cases

* Got action_sequence_to_logical_form working for simple cases

* Fix (most?) NLVR and WikiTables tests

* Renamed Executor to DomainLanguage

* Fix world test

* Add some documentation, fix pylint

* mypy

* Added docstrings, fixed docs

* Improve documentation, simplify some logic

* Removed the old type_declaration code entirely, better handled constants

* mypy

* PR feedback

* mypy...

* Actually fix mypy...

epwalsh committed:
Add CopyNet seq2seq model (#2237)

* add CopyNet seq2seq model

* add predictor

* fix for torch 1.0

* add version to deprecation warning

* Improve dataset reader docstring

* ensure tgt token indexers has SingleIdTokenIndexer

* make mypy happy

* fix serialized config

* get rid of warnings

* CopyMapField -> NamespaceSwappingField

* small code and doc improvements

* add copy token to vocab when model inits

* remove manual adding of copy token

vidurj committed:
Masked Flip (#2299)

Adds a `masked_flip` function that flips a tensor along the time dimension while respecting padding. This is useful for running an RNN in the backward direction along a padded batch. Eg: `masked_flip(lstm(masked_flip(batch, sequence_lengths))[0], sequence_lengths)`.

WrRan committed:
Check train/dev/test-file-path before process (#2308)

* check train/dev/test-file-path before process

* allow data_path to be a dir

schmmd committed:
Add raw prefix to avoid warnings from regex (#2310)

* Add raw prefix to avoid warnings from regex

* Update openai_transformer_byte_pair_indexer.py

DumbMachine committed:
Added links for some tutorials and organized the How-to alphabetically. (#2315)

Added links for Tutorials  for the following:
    Creating a Configuration File.
    Using a Debugger.
    Training and Using a Transformer ELMo.

Fixes #2284

generall committed:
Decoder class interfaces, initial scaffold

epwalsh committed:
Display default values in help message for allennlp command (#2323)

* add custom arg parser

* add overrides decorator

schmmd committed:
Fix backslash exceptions. (#2322)

epwalsh commented:
Thanks @generall! My one comment is that I think the `.forward()` should implement the entire forward loop, not just a single decoding step. What do you think?

Maybe the base decoder could implement the entire `.forward()` method, where it would call `.forward_step()` or something (which would be implemented by the subclasses). What do you think?

nelson-liu committed:
Skip Custom Highway LSTM tests, since they're broken (#2307)

Skipping these tests, since they're pretty much useless in their current state.

Have we decided if we're going to fix the custom extensions, or remove the module? I think there was talk of the latter, at some point?

apmoore1 committed:
Stacked Bi-Directional LSTM to be compatible with the Seq2Vec wrapper (#2318)

* Stacked Bi-directional LSTM works with Seq2Vec

* Doc String includes use_highway

* remove trailing spaces for pylint

DeNeutoy committed:
remove custom extensions (#2332)

fixes #2307 and fixes #1456

joelgrus committed:
add infer_and_cast (#2324)

* add infer_and_cast

* remove print statement + add comment

* address PR feedback

* pylint

generall commented:
Hi, I can see several reasons why `_forward_loop` should not be implemented in decoder:

1. `_forward_loop` does not contain all decoder code. It also presence in `take_step` function which is used inside BeamSearch
2. To implement `_forward_loop` functionality we need to have `_end_index` and `num_classes`. This values depend on `vocab`, and AFAIK we can't directly pass vocabulary to non-model class on `__init__`. In this case, we will have to invent additional initialization inside Seq2Seq2.
3. As I understand, it is good to have components as simple as possible. Since we don't want to re-implement BeamSearch and other tricks for better training in every decoder class.

Of course it is desirable to decompose BeamSearch and other training code out from Seq2Seq, but I think that 
this code should not be implemented inside Registrable component.

brendan-ai2 committed:
Evaluate on a token-weighted basis. (#2183)

- Allow models to return a `"batch_weight"` key that will be used to weight each batch's loss.
  - Per @matt-peters' suggestion.
  - Performed in Calypso: https://github.com/allenai/calypso/blob/master/calypso/train.py#L699
- Remove unused "loss_scale". This was never set in the training config, so it should be fairly safe.

epwalsh commented:
Gotcha. I think that's fair considering decoders could be pretty general, and we don't want to restrict ourselves to basic RNN decoders.

epwalsh commented on allennlp/modules/seq2seq_decoders/seq2seq2_decoder.py:
```diff
@@ -0,0 +1,24 @@
+    Decoder should implement function to perform a single decoding step.
+    """
+
+    def __init__(self, output_dim):
```
I don't think `output_dim` should be passed to `__init__`. Usually the decoder will output a `(batch_size, sequence_length)`, so `output_dim` would be `sequence_length`, but `sequence_length` is variable.

generall replied:
Since we agreed, that decoder should be separated from Vocabulary and all Embeddings, we can't decode actual output token id inside decoder. We can only output predicted target embedding which should have dimensionality `output_dim` (or as I renamed it later `decoding_dim`). It's also important to return embedding for `BeamSearch`, because it may chose different indexes based on random search policy.

epwalsh replied:
IMHO a `Seq2SeqDecoder` should be in charge of outputting an entire sequence, not just a single hidden state or prediction from one timestep, otherwise it should probably be named differently. I'm also not convinced that this abstraction for a "seq2step" is general enough. As is, it wouldn't exactly work for CopyNet: https://github.com/allenai/allennlp/blob/master/allennlp/models/encoder_decoders/copynet_seq2seq.py

generall replied:
Ok, good point on CopyNet. I will think about how to incorporate it into new decoder implementation. Agreed that current `Seq2SeqDecoder` should be called differently

epwalsh commented on allennlp/modules/seq2seq_decoders/seq2seq2_decoder.py:
```diff
@@ -0,0 +1,24 @@
+        super(Seq2SeqDecoder, self).__init__()
+        self._output_dim = output_dim
+
+    def get_output_dim(self) -> int:
```
Related to my other comment, we probably don't need this

generall replied:
Same here

schmmd committed:
Remove pylint ignores for backslashes. (#2331)

kl2806 committed:
QuaRel Domain Language (#2321)

* init quarel language

* add other predicates

* add and and quaval

* types and tests

* move domain lang out of test file

* quarel domain lang

* add comments

* tests

* tests

* rename vars

* mypy, pylint

* rename file

* test mutiple, no answer correct

* add doc

* Address style issues

* Check for inconsistent and comparison

* add whitespace

* reduce nesting

OyvindTafjord committed:
adjust call to lisp_to_nested_expression (#2347)

generall committed:
Done:

* Simple Seq2Seq decoder implemented
* Seq2Seq decoder integrated into SimpleSeq2Seq model
* Decoder accepts the whole decoding step history

ToDo:

* Write tests
* Implement Attention Transformer decoder as demonstration

generall commented:
My idea is to make a decoder in such a way that it can be implemented in this part.
![decoder](https://i.ibb.co/Vj0x3Hq/decoder.png).

We can also decompose all decoder-related functions out from SimpleSeq2Seq, but let's do it as separate part of this issue and focus first on decomposing of NN-related part of decoder.

WrRan committed:
Warn default value of min_padding_length (#2309)

* warn default-value of min_padding_length

* update warn type

* Update character_token_indexer_test.py

* add min_padding_length

* add min_padding_length

* add min_padding_length

* add min_padding_length

matt-gardner committed:
New NLVR language (#2319)

* Working on rewriting the NLVR language with DomainLanguage. Almost done...

* All execution tests pass; grammar induction probably isn't quite right yet

* Parsing tests now also pass

* Got higher-order functions working in Arithmetic

* negate_filter now works correctly in NlvrLanguage

* Working on converting NLVR code to use new API

* Mostly done, just need to fix processed data and saved models

* Fixed processed data

* Saving state

* Fix list/set bug

* Update fixtures

* Added partial list of sed commands for converting between old and new types

* Fix checklist info test

* Removed old code and all references to it, switched to new API

* Revert changes to train_fixtures.py

* pylint, mypy and docs

* Improving some docstrings

* Fix test

* Removed old TODO

* Fix problem introduced by rebase

* Update QuaRel test to use new function name

DNGros committed:
Fix very minor docstring typo in elmo.py (#2361)

For the Elmo module the docstring labels the type of scalar_mix_parameters as List[int] but the type hint labels it as List[float]. This confused me slightly at first. I am pretty sure float was intended.

joelgrus committed:
sanitize environment variables that can't be unicode encoded (#2357)

* sanitize environment variables that can't be unicode encoded

* fix mypy

matt-peters committed:
Add optional additional tokens to ELMo character indexer (#2364)

* Add optional additional tokens to ELMo character indexer

* fix mypy

apmoore1 committed:
Added StackedBiDirectional to list of encoders (#2339)

sai-prasanna committed:
Bump up pytorch-pretrained-bert to v0.4.0 (#2349)

nelson-liu committed:
Update create_a_configuration.md howto by removing old allennlp configure (#2330)

@joelgrus is there any way to access the old functionality of `allennlp configure`? or was it completely removed?

I've updated the `create_a_configuration` tutorial for the latter case, but happy to just edit run commands if the functionality is still around.

brendan-ai2 committed:
Add snippet for using `LanguageModelTokenEmbedder`. (#2359)

- Improves documentation for using transformer ELMo.
- A user requested an example of how to use transformer ELMo directly.
- https://github.com/allenai/allennlp/issues/2351

joelgrus committed:
trainer refactor (#2304)

* refactor

* work

* work

* wip

* address PR / CI feedback + more factoring

* pylint

* add toy multi-task trainer

* merge conflict for real

* abstract out early stopping into MetricTracker

* more refactoring

* move more functionality into tensorboard writer

* checkpointer

* change checkpointer interface

* try for backward compatibility

* remove some factoring

* address pr feedback, add gan example / test

* pylint

* fix docs

* fix whitespace

* add checkpointing to multitask trainer

* address PR feedback

* use rescale_gradients()

* fixes

* add base class from params + test

* fix pylint

epwalsh commented:
@generall thanks for this and sorry I've been slow to respond. I'll take a look first thing tomorrow

mise42 committed:
fix a typo (#2367)

change "experimenation" to "experimentation"

matt-peters committed:
Install package in dockerfile (#2305)

* Install package in dockerfile

* pip install editable at Michael's request

kl2806 committed:
Use f-string (#2370)

Commit:
Bag of words token embedder (#2365)

* added bow token embedder and tests

brendan-ai2 committed:
Disable requires_grad for Transformer ELMo parser. (#2336)

- Makes constituency_parser_transformer_elmo.jsonnet and constituency_parser_elmo.jsonnet comparable by default.

vidurj committed:
Sparse Gradient Clipping (#2312)

* not clamping sparse tensors

* removed trailing white space

* supporting sparse gradient clipping

* removed trailing white space

* ignore protected-access warning

* removed trailing whitespace

matt-gardner committed:
New WikiTables language (#2350)

* Working on moving WikiTables language to new API

* Logical form execution passes all of the original tests

* Got WikiTablesLanguage passing all tests

* Remove old executor / type declaration / world

* pylint, mypy, and docs

* Switch executor tests back to returning lists

* Address PR feedback

joelgrus committed:
pieces for multitask learning (#2369)

* pieces for multitask learning

* mypy

* add docs

* fix docs

* dashes to underscores

Commit:
changed log-to-console flags (#2381)

matt-gardner committed:
Make DomainLanguage handle executing action sequences directly, with side arguments (#2375)

* Make DomainLanguage handle executing action sequences directly, with side arguments

* mypy

* Address PR feedback

Commit:
Text classification JSON dataset reader (#2366)

* add text classification JSON dataset reader with tests

joelgrus committed:
move model to cuda in tests, add comment (#2384)

brendan-ai2 committed:
Rollback PR #2308 "Check train/dev/test-file-path" (#2386)

- This breaks the `MultiprocessDatasetReader` which relies on globs.
- Test to prevent regressing.
- The motivation behind PR #2308 is good, but we need to be a bit more sophisticated in how we perform the check. Perhaps `DatasetReader` should provide an optional `check_path` method that users can override?

HarshTrivedi committed:
Add support for extending Embeddings/Embedders with Extended Vocabulary. (#2374)

* Rough attempt for Embedder/Embedding extension.

* fix some mistakes.

* add tests for token-embedder and text-field-embedder extension.

* fix vocab_namespace usage in embedding.py

* update names and change some comments.

* update embedding tests.

* fix some typos.

* add more tests.

* update some comments.

* fix minor pylint issue.

* Implement extend_vocab for TokenCharactersEncoder.

* minor simplification.

* Update help text for --extend-vocab in fine-tune command.

* Shift location of model tests appropriately.

* Allow root Embedding in model to be extendable.

* Incorporate PR comments in embedding.py.

* Fix annotations.

* Add appropriate docstrings and minor cleanup.

* Resolve pylint complains.

* shift disable pytlint protected-access to top of tests.

* Add a test to increase coverage.

* minor update in TokenEmbedder docstrings.

* Remove a blank line.

* Add a blank line before Returns block in Embedding docstring.

joelgrus committed:
enable pickling for vocabulary (#2391)

* enable pickling for vocabulary

* pylint

HarshTrivedi committed:
Allow embedding extension to load from pre-trained embeddings file. (#2387)

* Rough attempt for Embedder/Embedding extension.

* fix some mistakes.

* add tests for token-embedder and text-field-embedder extension.

* fix vocab_namespace usage in embedding.py

* update names and change some comments.

* update embedding tests.

* fix some typos.

* add more tests.

* update some comments.

* fix minor pylint issue.

* Implement extend_vocab for TokenCharactersEncoder.

* minor simplification.

* Update help text for --extend-vocab in fine-tune command.

* Shift location of model tests appropriately.

* Allow root Embedding in model to be extendable.

* Incorporate PR comments in embedding.py.

* Fix annotations.

* Add appropriate docstrings and minor cleanup.

* Resolve pylint complains.

* shift disable pytlint protected-access to top of tests.

* Add a test to increase coverage.

* minor update in TokenEmbedder docstrings.

* Allow to pass pretrained_file in embedding extension (with tests).

* Remove a blank line.

* Add a blank line before Returns block in Embedding docstring.

* Fix pylint complains.

* Allow pretrained file to be passed in token_characters_encoder also.

* Fix pylint complains and update some comments.

* Test to ensure trained embeddings do not get overriden.

* PR feedback: update comments, fix annotation.

schmmd committed:
Remove unnecessary line (#2385)

Fixes https://github.com/allenai/allennlp/issues/2306

brendan-ai2 committed:
Remove scattering for multi-GPU training. (#2200)

- Instead just pull off a batch for each GPU.
- Enables increasing the effective batch size for `bidirectional_language_model.jsonnet` by 2x giving a 1.5x speedup.

matt-gardner committed:
Fix typing for python 3.7 (#2393)

* Fix typing for python 3.7

* Fix mypy, check sys.version_info instead of using try/catch

* Better way of checking version

bryant1410 committed:
Fix docstring for PretrainedBertIndexer's pretrained_model param (#2399)

`pretrained_model` is not an optional param, as stated in `PretrainedBertIndexer` docs, so I removed the "optional" part from it.

alasdairtran committed:
Ensure vocab is popped off params (#2409)

nelson-liu committed:
Remove outdated reference to custom extensions (#2401)

sodre committed:
Replace scripts with entry_points.console_scripts (#2232)

* Replace scripts with entry_points.console_scripts

  The recommended approach to create launch scripts is to use entry_points
  and console_scripts.

  xref: https://packaging.python.org/guides/distributing-packages-using-setuptools/#scripts

* Remove bin directory from Dockerfile, it is not needed

* Update references to bin/allennlp to just allennlp.

* Update argv in test.

joelgrus committed:
serialization in the tutorial (#2412)

* serialization in the tutorial

* add serialization to tutorial

* address PR feedback

* add disclaimer about programmatic generation

mahnerak committed:
Get rid of hardcoded Vocabulary class name. (#2418)

In case of inheriting Vocabulary, the method `from_params` calls the
same method of the registered class if `type` is given. If
`from_params` is not overridden (even if it is), the default
implementation calls hardcoded `Vocabulary.from_instances` and
`Vocabulary.from_files` which makes overriding these methods in
subclass unreachable.

DeNeutoy committed:
fix params duplicate bug (#2421)

HarshTrivedi committed:
Add AUC metric. (#2358)

* Add AUC metric.

* add masking.

* fix whitespace.

* fix docs.

* Incorporate PR feedback.

nelson-liu committed:
Fix typo when loading train state (#2425)

* Fix typo when loading train state

* Fix comment

ethanjperez committed:
Fixing BERT mask size (#2429)

* Fixing BERT mask size

Input tokens are truncated to <512 tokens long to fit into BERT. When this happens, the mask associated with that sample appears to not also be truncated. This edit is a fix for that issue, which can cause size mismatch errors downstream when the mask is used.

* update comment

matt-gardner committed:
Improve error message for "undocumented modules" (#2427)

brendan-ai2 committed:
Mark Atis Parser tests as flaky. Fix pylint. (#2430)

- Tests appear to pull down a DB file from S3.
- Spurious network failures are common, so mark as flaky to ameliorate the problem.
- Alternative: Bring DB file into fixtures directory. However, as we're trying to keep the size of the library down, this probably isn't advisable.
- Also disables some `pylint` warnings that hit with `numpy==1.16.0`.

yizhongw committed:
Optimizing the memory usage in `multi_head_self_attention` and `masked_softmax` (#2405)

* Optimizing the memory usage in `multi_head_self_attention` and `masked_softmax`.

Fixes #2185

* Fixes pylint error

* Fixes typo and change the attention dropout back to its original version.

yizhongw committed:
Support exponential moving average in the default trainer. (#2406)

* Support exponential moving average in the default trainer.

* Fixes pylint and type checker errors.

* Add doc string for the `ExponentialMovingAverage` module

* Add doc string for the `ExponentialMovingAverage` module

* add moving averages

* add docs

* Fixes pylint and doc.

* address pr feedback

HarshTrivedi committed:
Fix edgecase of potentially missing best-val-metrics in metrics.json on training recovery (#2352)

* Save and load best val metrics from training state.

* update trainer test for updated _restore_checkpoint.

* fix pylint, and minor fix.

* add tests.

* fix mypy issues.

* update test.

* fix a typo.

* Updated changes.

* Cleanup metric_tracker to track best_epoch directly.

* Shift appropriate logic in _restore_checkpoint.

* Update tests.

* Resolve mypy complains.

* Fix a typo from original refactor. Add a test.

* shift disable=protected-access to top of trainer_test.py

* Shift best_epoch and best_epoch_metrics out of _restore_checkpoint.

generall committed:
Rename Seq2SeqDecoder to DecoderCell

WrRan committed:
rename test case (#2441)

Before, in `allennlp/tests/modules/seq2seq_encoders/pass_through_encoder_test.py`:
```python
class TestStackedSelfAttention(AllenNlpTestCase)
    # ...
```

after:
```python
class TestPassThroughEncoder(AllenNlpTestCase):
    # ...
```

generall committed:
Attempt to decompose Seq2Seq models, make universal interface for simple_seq2sseq and CopyNet

WrRan committed:
Instantiate the class `Activation` properly when testing FeedForward. (#2443)

* `RELU` -> `Activation.by_name('relu')()`

* add some comment

epwalsh committed:
Generalize LR scheduler (#2345)

* start base scheduler

* fixes

* reorg

* add LearningRateScheduler base

* update custom LR schedulers

* fixes and update tests

* reorg tests

* fixes

* fix trainer

* fix docs

* fix trainer test

* make wrappers private

* improve docstring

* add momentum schedulers

* fix formatting for docs

* revert PyTorch LR scheduler behavior

WrRan committed:
[Feature Enhanced]Add FeedforwardEncoder for Sequences (#2444)

* add feedforward_encoder

* add mask & make mypy happy

* add doc

* mask on the output instead of the input

Hyperparticle committed:
Cleanup global logging after training (#2458)

WrRan committed:
Update `Running AllenNLP` in README.md (#2447)

Add subcommands: `configure`, `find-lr`:
```bash
$ allennlp
Run AllenNLP

optional arguments:
  -h, --help    show this help message and exit
  --version     show program's version number and exit

Commands:
  
    configure   Run the configuration wizard.
    train       Train a model
    evaluate    Evaluate the specified model + dataset
    predict     Use a trained model to make predictions.
    make-vocab  Create a vocabulary
    elmo        Create word vectors using a pretrained ELMo model.
    fine-tune   Continue training a model on a new dataset
    dry-run     Create a vocabulary, compute dataset statistics and other
                training utilities.
    test-install
                Run the unit tests.
    find-lr     Find a learning rate range.
```

matt-gardner committed:
Fix path issue for certain cases when using --include-package (#2464)

WrRan committed:
Make `mask` in `PassThroughEncoder` work (#2448)

The current `PassThroughEncoder` simply returns the input without taking into account the role of the parameter `mask`. I don't think this is a good idea for two reasons:
1. We should try our best to ensure that the same parameters of similar modules play the same role instead of ignoring it;
2. Using the parameter `mask` is more flexible. If we don't want to mask out any part, just pass in `None` (default value).

This PR is for it. What do you think?

WrRan committed:
More help info for `archive_surgery.py` (#2327)

* Fix bug in uniform_unit_scaling #2239 (#2273)

* Fix type annotation for .forward(...) in tutorial (#2122)

* Add a Contributions section to README.md (#2277)

* script for doing archive surgery (#2223)

* script for doing archive surgery

* simplify script

* Fix spelling in tutorial README (#2283)

* fix #2285 (#2286)

* Update the `find-lr` subcommand help text. (#2289)

* Update the elmo command help text.

* Update the find-lr subcommand help text.

* Add __repr__ to Vocabulary (#2293)

As it currently stands, the following is logged during training:

```
2019-01-06 10:46:21,832 - INFO - allennlp.common.from_params - instantiating class <class 'allennlp.model
s.language_model.LanguageModel'> from params {'bidirectional': False, 'contextualizer': {'bidirectional':
 False, 'dropout': 0.5, 'hidden_size': 200, 'input_size': 200, 'num_layers': 2, 'type': 'lstm'}, 'dropout
': 0.5, 'text_field_embedder': {'token_embedders': {'tokens': {'embedding_dim': 200, 'type': 'embedding'}
}}} and extras {'vocab': <allennlp.data.vocabulary.Vocabulary object at 0x7ff7811665f8>}
```

Note that the `Vocabulary` does not provide any useful information, since it doesn't have `__repr__` defined. This provides a fix.

* Update the base image in the Dockerfiles. (#2298)

* Don't deprecate bidirectional-language-model name (#2297)

* bump version number to v0.8.1

* Bump version numbers to v0.8.2-unreleased

* Turn BidirectionalLM into a more-general LanguageModel class (#2264)

Fixes https://github.com/allenai/allennlp/issues/2255

This PR replaces the `BidirectionalLM` class with a more-general `LanguageModel` that can be used in either the unidirectional/forward setting or the bidirectional setting.

It also accordingly replaces the `BidirectionalLanguageModelTokenEmbedder` with a `LanguageModelTokenEmbedder`.

Also fixes bug in the experiment_unsampled.jsonnet config that was preventing a test from actually being unsampled.

TODO:

- [x] test the unidirectional case
- [x] properly deprecate `BidirectionalLM` and `BidirectionalLanguageModelTokenEmbedder` 
- [x] check docs for accuracy
- [x] fix user-facing training configs

* more help info

* typo fix

* add option '--inplace', '--force'

* clearer help text

Meelfy committed:
Ten times faster than before in GPU get_best_span of bidaf (#2465)

* faster(GPU) get_best_span in bidaf

* reanme

* chang the way of mask: from multipilcation to addition

* Improve names and comments

bryant1410 committed:
Add support to kwargs in TimeDistributed (#2439)

* Add support to kwargs in TimeDistributed

* Add a TODO in the named arg and kwarg TD test

* Fix PR comments

* Remove unused import

* Fix a PR comment: Iterable -> List

* Add missing @overrides

* Add tests

* Add missing forward methods pylint warning suppression

yizhongw committed:
Fixes unnecessary parameter clone in moving average. (#2468)

yizhongw committed:
Add QaNet model (#2446)

* Add max length limit for passage and question in SQuAD reader.

* Add QaNet model.

* fixes the squad reader and adds doc.

* Move `get_best_span()` function out of bidaf.

* Update the docstring of QANet and BiDAF

* Move `ResidualWithLayerDropout` to a separate module file.

* Update the docstring and test cases for the length limits in squad reader.

* Keep the old `get_best_span` function in `bidaf.py`.

* Add docstring for `get_best_span` function.

* Separate test case for the `get_best_span` function.

* Fixes docs.

* Update the training configuration file.

* ignores pylint error.

* add docs for layer dropout.

* fixes docs.

* Remove the unsqueeze()

nelson-liu committed:
Fix CORS error for react in config explorer (#2476)

WrRan committed:
[Feature Enhanced] Support sentence pair in BERT (#2279)

* support sentence pair in BERT

* support sentence pair in BERT

* fix bug: `get_keys` in WordpieceIndexer

* make pylint happy

* add doc & more efficient implementation

* make pylint and mypy happy

Commit:
make bag_of_word_counts token embedder ignore padding and UNK tokens (#2432)

* added patch to ignore padding and UNK tokens

schmmd committed:
WIP: Make warnings errors and filter library warnings from pytest (#2479)

HarshTrivedi committed:
Add support for pretrained embedding extension in fine-tuning. (#2395)

* First attempt to allow pretrained embedding extension in fine-tune.

* Fix some typos, fix mypy and pylint.

* Revert changes in Train and Archival.

* Allow mapping to be explicitly passed in fine-tune, instead of getting it from  archive.

* Use model_path to pretrained-file path mapping.

* Remove unnessary code.

* Remove tests for unnecessary code.

* Correct typos.

* Fix minor mistakes.

* Fix pylint issues.

* Change error to warning in embedding extension.

* Add test to check mapping works.

* Remove test for the remove code.

* Fix mypy and pylint complains.

* Add more tests.

* Fix a typo.

* Make sure embedding extension is no-op when required.

* Use fixture in test.

* Update docstring and helper message.

* Fix file availability bug and cleanup.

* Update docstring.

* Minor cleanup.

* Add missing fixture file.

* Fix pylint issues.

* Incorporate PR feedback.

* Fix argument call.

* Update test.

* flatten logic

* fix none + check for 'extend_vocab' attribute

* remove unused import

* Fix condition typo.

matt-gardner committed:
Mark QaNet test as flaky (#2481)

* Mark QaNet test as flaky

* Import the decorator :facepalm:

apmoore1 committed:
Variational dropout in Augmented LSTM (#2344)

This pull requests fixes #2320 I have done point 2 on the list within the issue. Apart from the points made in the issue I have also added more detail to the doc strings in the [Stacked Bi-Directional LSTM](https://github.com/allenai/allennlp/blob/master/allennlp/modules/stacked_bidirectional_lstm.py) such as the corrected shape of the `final_states` returned form the forward pass and the correct shape required for the `initial_state` argument for the forward pass. I have also included the returned Type of the forward pass for the [Stacked Bi-Directional LSTM](https://github.com/allenai/allennlp/blob/master/allennlp/modules/stacked_bidirectional_lstm.py).

Screen shot of the changed doc strings for the augmented LSTM is below: Only change is the wording of the `recurrent_dropout_probability` argument within the constructor:
![augmented doc string](https://user-images.githubusercontent.com/13574854/51051958-34cf7180-15cd-11e9-9a48-7bbb039ae504.png)

Screen shots of the changed doc strings for the Stacked Bi-Directional LSTM are below: Changes are those stated above with regards to the forward pass and the wording of the `recurrent_dropout_probability` and `layer_dropout_probability` within the constructor as well as the text of the constructor:
![stacked constructor doc string](https://user-images.githubusercontent.com/13574854/51051945-2719ec00-15cd-11e9-9809-4a603a8e1a66.png)
![stacked forward doc string](https://user-images.githubusercontent.com/13574854/51051952-2d0fcd00-15cd-11e9-906b-7a1cf3ab7b25.png)

generall committed:
decompose simple seq2seq model, fix tests

generall commented:
Hi, I have decomposed decoder from seq2seq model. Right now it consists of:

* base class `SeqDecoder`
* simple decoder implementation `SimpleSeqDecoder`
* decoder cell, which contains only decoder NN layers `SimpleDecoderCell` with base class `DecoderCell`

I had some issues during my attempts to use `SeqDecoder` in copynet. I noticed that it has it's own dataset reader and I wonder if it is possible to somehow move it's preprocessing inside copynet. In this case it will be possible to implement copynet decomposed `SeqDecoder`.

HarshTrivedi committed:
Add support to transfer submodules, and in different modes. (#2471)

* Add method to find module path from module in model.

* Add method to add prevent_regexes in initializer.

* Add TransferModule helper.

* Add examples and tests to use TransferModules class.

* Revert previous changes.

* Add from_pretrained_params and tests.

* Use from_pretrained_params for modules in which from_params was overrided.

* Fix typos, pylint and mypy issues.

* Fix some more typos.

* Handle some edge cases.

* Add docstring to from_pretrained_params.

* Fix docstring syntax.

* Add return type annotation to from_pretrained_params.

* Correct typos in docstring.

* Fix mistake.

* Add test to assure type consistency of loaded module.

* Shift back from_pretrained_params to FromParams and fix mypy.

* Revert unnecessary code changes.

* Revert more unnecessary changes.

* Simpify implementation according to pr feedback.

* Fix edgecase in the condition; add comment.

* Shift comment above the condition.

* Shift load module to archival.

* Remove comment.

* Change module_path argument to path.

* Fix mypy.

* Fix docstring.

epwalsh committed:
Add momentum schedulers to trainer (#2469)

* add momentum scheduler to trainer

* add test for momentum scheduler in trainer

joelgrus committed:
upgrade pytorch-pretrained-bert (#2495)

matt-peters committed:
Less restrictive token characters indexer (#2494)

* Less restrictive token characters indexer

* fix tests

* add clarifying comment

* mypy

* add tests

* pylint

matt-gardner committed:
Revert "Less restrictive token characters indexer (#2494)" (#2497)

This reverts commit 89056f1e5063a3b10ab9230d8d6d0a43f5657fd0.

DeNeutoy committed:
remove notebooks and juptyer dependency (#2499)

HarshTrivedi committed:
Add necessary implicit embedding extension for transfer-modules api and vocab extension (#2431)

* Fix loading of fine-tuned + extended model.

* Refactor comments and remove unnecessary line.

* Assure transferred embedding modules are extended when required.

* Make embedding extension be no-op when extension isn't required.

* Don't allow embedding extension to assume "tokens" namespace.

* Test embedding extension to be no-op when appropriate.

* Add more tests.

* Fix a typo.

* update comments.

* Fix typos in test and add some comments.

* Remove duplicate condition.

* Change no namespace warning to info log level.

* Change no-op condition for extend_vocab call.

* Raise error if vocab or namespace is surely incorrect.

* Fix pylint.

* Update tests.

* Prevent incorrect vocab namespace assignment.

* Working tweaks in messages.

* Wording tweak in comment.

HarshTrivedi committed:
Allow vocab+embedding extension in evaluate command. (#2501)

* Allow vocab+embedding extension in evaluate.

* Fix typo.

* make extend_embedder_vocab call conditional.

* Add test.

* Update evaluate command help.

HarshTrivedi committed:
Remove extended_vocab argument from extend_embedder_vocab API. (#2505)

* Remove extended_vocab argument from extend_embedder_vocab api.

* Fix test call.

WrRan committed:
fix typo (#2508)

generall commented:
Hi again @epwalsh , would you have a time to review my changes?

epwalsh commented:
Hey @generall, sorry about the late response, this completely fell off of my radar. Honestly I don't think I'll have much time to help out with this anytime soon, but it looks like you've got a pretty good start, so I'd suggest continuing this on your branch, and making a WIP (work in progress) pull request to AllenNLP (the main repo). I'd be happy to offer suggestions before you make an official PR if you'd like.

generall committed:
fix comments and add dimensions check

saujasv committed:
Adding bag_of_embeddings as an alternate name for the boe encoder (Issue #2268) (#2521)

joelgrus committed:
upgrade huggingface pretrained bert (#2524)

schmmd committed:
bump version number to v0.8.2

schmmd committed:
Bump version numbers to v0.8.3-unreleased

matt-gardner committed:
Restore tensorboard epoch metrics to pre-refactoring behavior (#2532)

* Make "epoch_metrics" prefix consistent

* Restore pre-refactor metric behavior

DeNeutoy committed:
make archival take an optional output path (#2510)

brendan-ai2 committed:
Propose a deprecation policy. (#2424)

- Examples of when backward compatibility is likely to be an issue.
- Suggestions for ensuring compatibility.
- Mechanics for deprecating.

brendan-ai2 committed:
Clarify `data_parallel` implementation. (#2488)

- It appears we're unnecessarily replicating the model to the source device, but this is not the case.
- Silence some deprecation warnings that were missed previously.

generall committed:
Make decoder state more general

generall committed:
renames + minor fixes + predictor test

matt-gardner committed:
Scope DeprecationWarning errors to just allennlp-internal stuff (#2549)

kyleclo committed:
missing `=overrides` argument when instantiate Params despite a second time (#2553)

generall committed:
Merge branch 'master' into decoder-updates

generall committed:
update docstring

generall committed:
Merge branch 'decoder-updates' of github.com:generall/allennlp into decoder-updates

generall committed:
Registrable SeqDecoder

generall committed:
Remove decoder_cell from SeqDecoder, rename SimpleSeqDecoder -> RnnSeqDecoder
